File: TwitterElectron/TwitterElectron/RendererProcess/MainForm.cs
using System;
using Bridge;
using TwitterElectron.Twitter;
using TwitterElectron.Twitter.API;
using static Retyped.dom;
using static Retyped.electron;
using static Retyped.node;
using static Retyped.jquery;

namespace TwitterElectron.RendererProcess
{
    public static class MainForm
    {
        private const int MaxTweetsOnPage = 20;
        private const int NotificationBufferTimeSec = 20;

        private const string LightThemeCss = "../Assets/Styles/light.css";
        private const string DarkThemeCss = "../Assets/Styles/dark.css";

        [Init(InitPosition.Top)]
        public static void InitGlobals()
        {
            // Init global variables (modules):
            var Electron = (Electron.AllElectron)require.Self("electron");
            var jQuery = require.Self("jquery");
        }

        [Template("Electron")]
        public static Electron.AllElectron Electron;

        private static TwitterListener _listener;

        private static DateTime? _lastNotificationDate;

        public static void Main()
        {
            // Configure handlers for IPC commands:
            ConfigureIPC();

            // Configure handlers for HtmlElement events:
            ConfigureEventHandlers();

            // Set Theme:
            ToggleTheme();
        }

        private static void ConfigureEventHandlers()
        {
            jQuery.select(".play").on("click", (e, args) =>
            {
                Electron.ipcRenderer.send(Constants.IPC.StartCapture);
                return null;
            });

            jQuery.select(".pause").on("click", (e, args) =>
            {
                Electron.ipcRenderer.send(Constants.IPC.StopCapture);
                return null;
            });
        }

        private static void ConfigureIPC()
        {
            Electron.ipcRenderer.on(Constants.IPC.StartCapture, () =>
            {
                jQuery.select("#placeholder").hide();
                jQuery.select(".play").hide();
                jQuery.select(".pause").show();

                _listener = InitListener();
                if (_listener != null)
                {
                    _listener.Start();
                }
                else
                {
                    // Can't start capturing due to some error (no filter, no credentials)
                    Electron.ipcRenderer.send(Constants.IPC.StopCapture);
                }
            });

            Electron.ipcRenderer.on(Constants.IPC.StopCapture, () =>
            {
                jQuery.select(".pause").hide();
                jQuery.select(".play").show();

                _listener?.Stop();
            });

            Electron.ipcRenderer.on(Constants.IPC.ClearCapture, () =>
            {
                jQuery.select("#capturedItemsDiv").html("");
                jQuery.select("#placeholder").show();
            });

            Electron.ipcRenderer.on(Constants.IPC.ToggleTheme, new Action<Electron.Event>(ev =>
            {
                ToggleTheme();
            }));
        }

        private static TwitterListener InitListener()
        {
            // Get credentials from the main process:
            var credentials = (TwitterCredentials) Electron.ipcRenderer.sendSync(Constants.IPC.GetCredentialsSync);

            // Check credentials:
            if (credentials == null ||
                string.IsNullOrEmpty(credentials.ApiKey) ||
                string.IsNullOrEmpty(credentials.ApiSecret) ||
                string.IsNullOrEmpty(credentials.AccessToken) ||
                string.IsNullOrEmpty(credentials.AccessTokenSecret))
            {
                alert("Please specify API keys and Access tokens before starting.");

                return null;
            }

            // Check filter value:
            var filter = (string) jQuery.select("#captureFilterInput").val();
            if (string.IsNullOrEmpty(filter))
            {
                alert("Please specify filter value.");

                return null;
            }

            // Create Twitter stream listener:
            var listener = new TwitterListener(credentials, filter);

            // Configure handlers for the created listener events:
            listener.OnReceived += (sender, tweet) =>
            {
                AddTweetToPage(tweet);

                // Notify about the obtained tweet:
                var notificationEnabled = jQuery.select("#notificationEnabledCheckbox").@is(":checked");
                if (notificationEnabled)
                {
                    // Use delay to avoid creating too many notifications:
                    if (_lastNotificationDate == null ||
                        (DateTime.UtcNow - _lastNotificationDate.Value).TotalSeconds > NotificationBufferTimeSec)
                    {
                        _lastNotificationDate = DateTime.UtcNow;
                        CreateNotification(tweet);
                    }
                }
                else
                {
                    _lastNotificationDate = null;
                }
            };

            listener.OnError += (sender, err) =>
            {
                // Stop capturing on error:
                Electron.ipcRenderer.send(Constants.IPC.StopCapture);
                alert($"Error: {err}");
            };

            return listener;
        }

        private static void ToggleTheme()
        {
            var lightThemeLink = jQuery.select($"link[href='{LightThemeCss}']");
            var darkThemeLink = jQuery.select($"link[href='{DarkThemeCss}']");

            var newTheme = lightThemeLink.length == 0
                ? LightThemeCss
                : DarkThemeCss;

            if (lightThemeLink.length == 0)
            {
                darkThemeLink.remove();
            }
            else if (darkThemeLink.length == 0)
            {
                lightThemeLink.remove();
            }

            jQuery.select("head").append($"<link rel=\"stylesheet\" href=\"{newTheme}\" >");
        }

        private static void CreateNotification(Tweet tweet)
        {
            var notifTitle = tweet.user.name + " is tweeting..";

            var notifOpts = ObjectLiteral.Create<NotificationOptions>();
            notifOpts.body = tweet.text;
            notifOpts.icon = tweet.user.profile_image_url;

            var notif = new Notification(notifTitle, notifOpts);
            notif.onclick = notifEv =>
            {
                var tweetUrl = $"https://twitter.com/{tweet.user.screen_name}/status/{tweet.id_str}";

                Electron.shell.openExternal(tweetUrl);

                return null;
            };
        }

        private static void AddTweetToPage(Tweet tweet)
        {
            var div = new HTMLDivElement
            {
                className = "tweet-card animated slideInRight"
            };

            div.ondblclick += e =>
            {
                var tweetUrl = $"https://twitter.com/{tweet.user.screen_name}/status/{tweet.id_str}";
                Electron.shell.openExternal(tweetUrl);

                return null;
            };

            var img = new HTMLImageElement
            {
                className = "avatar",
                src = tweet.user.profile_image_url
            };

            var nameDiv = new HTMLDivElement
            {
                className = "username",
                innerHTML = tweet.user.name + "<span class='istweeting'> is tweeting...</span>"
            };

            var textDiv = new HTMLDivElement
            {
                className = "tweet-text",
                innerHTML = tweet.text
            };

            var tweetContent = new HTMLDivElement {className = "tweet-content"};
            tweetContent.appendChild(nameDiv);
            tweetContent.appendChild(textDiv);

            div.appendChild(img);
            div.appendChild(tweetContent);

            var capturedItemsDiv = jQuery.select("#capturedItemsDiv");
            var capturedItems = capturedItemsDiv.children();

            if (capturedItems.length > 0)
            {
                if (capturedItems.length >= MaxTweetsOnPage)
                {
                    capturedItems[MaxTweetsOnPage - 1].remove();
                }
            }

            capturedItemsDiv.prepend(div);
        }
    }
}

File: TwitterElectron/TwitterElectron/App/MainForm.js
// Init global variables (modules):
var Electron = require("electron");
var jQuery = require("jquery");

Bridge.assembly("TwitterElectron", function ($asm, globals) {
    "use strict";

    Bridge.define("TwitterElectron.RendererProcess.MainForm", {
        main: function Main () {
            // Configure handlers for IPC commands:
            TwitterElectron.RendererProcess.MainForm.ConfigureIPC();

            // Configure handlers for HtmlElement events:
            TwitterElectron.RendererProcess.MainForm.ConfigureEventHandlers();

            // Set Theme:
            TwitterElectron.RendererProcess.MainForm.ToggleTheme();
        },
        statics: {
            fields: {
                MaxTweetsOnPage: 0,
                NotificationBufferTimeSec: 0,
                LightThemeCss: null,
                DarkThemeCss: null,
                Electron: null,
                _listener: null,
                _lastNotificationDate: null
            },
            ctors: {
                init: function () {
                    this.MaxTweetsOnPage = 20;
                    this.NotificationBufferTimeSec = 20;
                    this.LightThemeCss = "../Assets/Styles/light.css";
                    this.DarkThemeCss = "../Assets/Styles/dark.css";
                }
            },
            methods: {
                ConfigureEventHandlers: function () {
                    jQuery(".play").on("click", $asm.$.TwitterElectron.RendererProcess.MainForm.f1);

                    jQuery(".pause").on("click", $asm.$.TwitterElectron.RendererProcess.MainForm.f2);
                },
                ConfigureIPC: function () {
                    Electron.ipcRenderer.on("cmd-start-capture", $asm.$.TwitterElectron.RendererProcess.MainForm.f3);

                    Electron.ipcRenderer.on("cmd-stop-capture", $asm.$.TwitterElectron.RendererProcess.MainForm.f4);

                    Electron.ipcRenderer.on("cmd-clear-capture", $asm.$.TwitterElectron.RendererProcess.MainForm.f5);

                    Electron.ipcRenderer.on("cmd-toggle-theme", $asm.$.TwitterElectron.RendererProcess.MainForm.f6);
                },
                InitListener: function () {
                    // Get credentials from the main process:
                    var credentials = Electron.ipcRenderer.sendSync("cmd-get-credentials-sync");

                    // Check credentials:
                    if (credentials == null || System.String.isNullOrEmpty(credentials.ApiKey) || System.String.isNullOrEmpty(credentials.ApiSecret) || System.String.isNullOrEmpty(credentials.AccessToken) || System.String.isNullOrEmpty(credentials.AccessTokenSecret)) {
                        alert("Please specify API keys and Access tokens before starting.");

                        return null;
                    }

                    // Check filter value:
                    var filter = Bridge.cast(jQuery("#captureFilterInput").val(), System.String);
                    if (System.String.isNullOrEmpty(filter)) {
                        alert("Please specify filter value.");

                        return null;
                    }

                    // Create Twitter stream listener:
                    var listener = new TwitterElectron.Twitter.TwitterListener(credentials, filter);

                    // Configure handlers for the created listener events:
                    listener.addOnReceived($asm.$.TwitterElectron.RendererProcess.MainForm.f7);

                    listener.addOnError($asm.$.TwitterElectron.RendererProcess.MainForm.f8);

                    return listener;
                },
                ToggleTheme: function () {
                    var lightThemeLink = jQuery(System.String.format("link[href='{0}']", TwitterElectron.RendererProcess.MainForm.LightThemeCss));
                    var darkThemeLink = jQuery(System.String.format("link[href='{0}']", TwitterElectron.RendererProcess.MainForm.DarkThemeCss));

                    var newTheme = lightThemeLink.length === 0 ? TwitterElectron.RendererProcess.MainForm.LightThemeCss : TwitterElectron.RendererProcess.MainForm.DarkThemeCss;

                    if (lightThemeLink.length === 0) {
                        darkThemeLink.remove();
                    } else if (darkThemeLink.length === 0) {
                        lightThemeLink.remove();
                    }

                    jQuery("head").append(System.String.format("<link rel=\"stylesheet\" href=\"{0}\" >", newTheme));
                },
                CreateNotification: function (tweet) {
                    var notifTitle = (tweet.user.name || "") + " is tweeting..";

                    var notifOpts = { };
                    notifOpts.body = tweet.text;
                    notifOpts.icon = tweet.user.profile_image_url;

                    var notif = new Notification(notifTitle, notifOpts);
                    notif.onclick = function (notifEv) {
                        var tweetUrl = System.String.format("https://twitter.com/{0}/status/{1}", tweet.user.screen_name, tweet.id_str);

                        Electron.shell.openExternal(tweetUrl);

                        return null;
                    };
                },
                AddTweetToPage: function (tweet) {
                    var $t;
                    var div = ($t = document.createElement("div"), $t.className = "tweet-card animated slideInRight", $t);

                    div.ondblclick = Bridge.fn.combine(div.ondblclick, function (e) {
                        var tweetUrl = System.String.format("https://twitter.com/{0}/status/{1}", tweet.user.screen_name, tweet.id_str);
                        Electron.shell.openExternal(tweetUrl);

                        return null;
                    });

                    var img = ($t = document.createElement("img"), $t.className = "avatar", $t.src = tweet.user.profile_image_url, $t);

                    var nameDiv = ($t = document.createElement("div"), $t.className = "username", $t.innerHTML = (tweet.user.name || "") + "<span class='istweeting'> is tweeting...</span>", $t);

                    var textDiv = ($t = document.createElement("div"), $t.className = "tweet-text", $t.innerHTML = tweet.text, $t);

                    var tweetContent = ($t = document.createElement("div"), $t.className = "tweet-content", $t);
                    tweetContent.appendChild(nameDiv);
                    tweetContent.appendChild(textDiv);

                    div.appendChild(img);
                    div.appendChild(tweetContent);

                    var capturedItemsDiv = jQuery("#capturedItemsDiv");
                    var capturedItems = capturedItemsDiv.children();

                    if (capturedItems.length > 0) {
                        if (capturedItems.length >= TwitterElectron.RendererProcess.MainForm.MaxTweetsOnPage) {
                            capturedItems[19].remove();
                        }
                    }

                    capturedItemsDiv.prepend(div);
                }
            }
        }
    });

    Bridge.ns("TwitterElectron.RendererProcess.MainForm", $asm.$);

    Bridge.apply($asm.$.TwitterElectron.RendererProcess.MainForm, {
        f1: function (e, args) {
            Electron.ipcRenderer.send("cmd-start-capture");
            return null;
        },
        f2: function (e, args) {
            Electron.ipcRenderer.send("cmd-stop-capture");
            return null;
        },
        f3: function () {
            jQuery("#placeholder").hide();
            jQuery(".play").hide();
            jQuery(".pause").show();

            TwitterElectron.RendererProcess.MainForm._listener = TwitterElectron.RendererProcess.MainForm.InitListener();
            if (TwitterElectron.RendererProcess.MainForm._listener != null) {
                TwitterElectron.RendererProcess.MainForm._listener.Start();
            } else {
                // Can't start capturing due to some error (no filter, no credentials)
                Electron.ipcRenderer.send("cmd-stop-capture");
            }
        },
        f4: function () {
            jQuery(".pause").hide();
            jQuery(".play").show();

            TwitterElectron.RendererProcess.MainForm._listener != null ? TwitterElectron.RendererProcess.MainForm._listener.Stop() : null;
        },
        f5: function () {
            jQuery("#capturedItemsDiv").html("");
            jQuery("#placeholder").show();
        },
        f6: function (ev) {
            TwitterElectron.RendererProcess.MainForm.ToggleTheme();
        },
        f7: function (sender, tweet) {
            TwitterElectron.RendererProcess.MainForm.AddTweetToPage(tweet);

            // Notify about the obtained tweet:
            var notificationEnabled = jQuery("#notificationEnabledCheckbox").is(":checked");
            if (notificationEnabled) {
                // Use delay to avoid creating too many notifications:
                if (Bridge.equals(TwitterElectron.RendererProcess.MainForm._lastNotificationDate, null) || (System.DateTime.subdd(System.DateTime.getUtcNow(), System.Nullable.getValue(TwitterElectron.RendererProcess.MainForm._lastNotificationDate))).getTotalSeconds() > TwitterElectron.RendererProcess.MainForm.NotificationBufferTimeSec) {
                    TwitterElectron.RendererProcess.MainForm._lastNotificationDate = System.DateTime.getUtcNow();
                    TwitterElectron.RendererProcess.MainForm.CreateNotification(tweet);
                }
            } else {
                TwitterElectron.RendererProcess.MainForm._lastNotificationDate = null;
            }
        },
        f8: function (sender, err) {
            // Stop capturing on error:
            Electron.ipcRenderer.send("cmd-stop-capture");
            alert(System.String.format("Error: {0}", err));
        }
    });
});
